spi/zynq7000-xspi/libzynq7000-xspi.c

```diff
@@ -0,0 +1,308 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Zynq-7000 AXI SPI controller (master mode only)
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Lukasz Kosinski, Aleksander Kaminski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <spi.h>
+#include <stdint.h>
+
+#include <sys/interrupt.h>
+#include <sys/minmax.h>
+#include <sys/mman.h>
+#include <sys/platform.h>
+#include <sys/threads.h>
+#include <sys/types.h>
+
+#include <phoenix/arch/zynq7000.h>
+#include <board_config.h>
+
+#include "libzynq7000-xspi.h"
+
+
+/* SPI registers */
+#define SPI_SRR   16 /* Software reset register */
+#define SPI_CR    24 /* Control register */
+#define SPI_SR    25 /* Status register */
+#define SPI_DTR   26 /* Data transfer register */
+#define SPI_DRR   27 /* Data receive register */
+#define SPI_SSR   28 /* Slave select register */
+#define SPI_TFIFO 29 /* Trasnmit fifo ocupancy register */
+#define SPI_RFIFO 30 /* Receive fifo ocupancy register */
+#define SPI_DGIER 7  /* Global interrupt enable register */
+#define SPI_ISR   8  /* Interrupt status register */
+#define SPI_IER   10 /* Interrupt enable register */
+
+/* SPI definitions */
+#define SPI_FIFO_SIZE 256 /* SPI TX and RX FIFO size */
+
+
+typedef struct {
+	const unsigned int id;   /* SPI controller ID */
+	const unsigned int irq;  /* SPI controller IRQ */
+	const addr_t paddr;      /* SPI controller base physical address */
+	volatile uint32_t *base; /* SPI registers base address */
+	handle_t lock;           /* SPI IRQ mutex */
+	handle_t cond;           /* SPI IRQ cond */
+	handle_t inth;           /* SPI IRQ handle */
+} spi_t;
+
+
+/* SPI controllers */
+static spi_t devs[] = {
+	{
+		.id = 0,             /* SPI0 */
+		.irq = 64,           /* SPI0 IRQ */
+		.paddr = 0x41e00000, /* SPI0 base physical address */
+	}
+};
+
+
+static void spi_dmb(void)
+{
+	__asm__ volatile ("dmb");
+}
+
+
+static spi_t *spi_get(unsigned int dev)
+{
+	if (dev >= sizeof(devs) / sizeof(devs[0])) {
+		return NULL;
+	}
+
+	return &devs[dev];
+}
+
+
+static int spi_isr(unsigned int n, void *arg)
+{
+	spi_t *spi = arg;
+
+	/* TX FIFO empty */
+	if (*(spi->base + SPI_ISR) & (1 << 2)) {
+		return 1;
+	}
+
+	/* Should never happen */
+	return -1;
```

Comment:
Why `-1` not `0`?

Reply:
Because: https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/0fe0bd19853c3728c649c79044c250bb9653f290/proc/userintr.c#L74

Condition for wakeup is ret >= 0